rust/yarp/src/lib.rs

```diff
@@ -0,0 +1,254 @@
+        Self(unsafe { (*(self.0)).node.next.cast::<yp_diagnostic_t>() })
+    }
+}
+
+impl ListNode<Self> for Comment {
+    fn is_null(&self) -> bool {
+        self.0.is_null()
+    }
+
+    fn next(&self) -> Self {
+        Self(unsafe { (*(self.0)).node.next.cast::<yp_comment_t>() })
+    }
+}
+
+/// An iterator that can be used to iterate over a linked list.
+pub struct ListIter<T: ListNode<T> + Copy> {
+    current: T
+}
+
+impl<T: ListNode<T> + Copy> Iterator for ListIter<T> {
+    type Item = T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.current.is_null() {
+            None
+        } else {
+            let current = self.current;
+            self.current = self.current.next();
+            Some(current)
+        }
+    }
+}
+
+/// The result of parsing a source string.
+pub struct ParseResult {
+    source: CString,
+    parser: NonNull<yp_parser_t>,
+    node: NonNull<yp_node_t>
+}
+
+impl ParseResult {
+    /// Returns the source string that was parsed.
+    /// 
+    /// # Panics
+    /// 
+    /// Panics if the source string is not able to be converted into a `&str`.
+    #[must_use]
+    pub fn source(&self) -> String {
+        self.source.to_str().unwrap().to_string()
+    }
+
+    /// Returns an iterator that can be used to iterate over the errors in the
+    /// parse result.
+    #[must_use]
+    pub fn errors_iter(&self) -> ListIter<Diagnostic> {
+        unsafe {
+            let list = &mut (*self.parser.as_ptr()).error_list;
+            ListIter { current: Diagnostic(list.head.cast::<yp_diagnostic_t>()) }
+        }
+    }
+
+    /// Returns an iterator that can be used to iterate over the warnings in the
+    /// parse result.
+    #[must_use]
+    pub fn warnings_iter(&self) -> ListIter<Diagnostic> {
+        unsafe {
+            let list = &mut (*self.parser.as_ptr()).warning_list;
+            ListIter { current: Diagnostic(list.head.cast::<yp_diagnostic_t>()) }
+        }
+    }
+
+    /// Returns an iterator that can be used to iterate over the comments in the
+    /// parse result.
+    #[must_use]
+    pub fn comments_iter(&self) -> ListIter<Comment> {
+        unsafe {
+            let list = &mut (*self.parser.as_ptr()).comment_list;
+            ListIter { current: Comment(list.head.cast::<yp_comment_t>()) }
+        }
+    }
+
+    /// Returns the root node of the parse result.
+    #[must_use]
+    pub fn node(&self) -> Node {
+        Node::new(self.node.as_ptr())
+    }
+}
+
+impl Drop for ParseResult {
+    fn drop(&mut self) {
+        unsafe {
+            yp_node_destroy(self.parser.as_ptr(), self.node.as_ptr());
+
+            // TODO: Why isn't this working?
+            // yp_parser_free(self.parser.as_ptr());
+        }
+    }
+}
+
+/// Parses the given source string and returns a parse result.
+/// 
+/// # Panics
+/// 
+/// Panics if the source string is not able to be converted into a `CString`.
+/// 
+#[must_use]
+pub fn parse(source: String) -> ParseResult {
```

Comment:
I would expect a function called `parse` to accept a borrow of its input: either `&str` or `&[u8]`.